### ISSUE TITLE
Impossible to alter column type that has skipping index on it (LC(String) to (String))

### ISSUE THREAD

**Issue by fhmcn** [ClickHouse/ClickHouse#17632]
**Issue**
while upgrading from 20.3.11.97 to 20.8.7.15-2 version. 
service startup failed due to invalid codec errors.

Error:
```
SQL Error [36]: ClickHouse exception, code: 36, host: 127.0.0.1, port: 54669; Code: 36, e.displayText() = DB::Exception: It does not make sense to have codec NONE along with other compression codecs: NONE, ZSTD(1). (Note: you can enable setting 'allow_suspicious_codecs' to skip this check). (version 20.8.7.15 (official build))
```

set allow_suspicious_codecs=1 in user config and restarted service. server failed with the below error.
```
Received exception from server (version 20.8.7):
Code: 36. DB::Exception: Received from localhost:9000. DB::Exception: Codec Gorilla is not applicable for LowCardinality(String) because the data type is not of fixed size. 
```

changing attach table SQL files did not help, it created confusion between local and ZK versions. 

**How to reproduce**
* Which ClickHouse server versions are incompatible
`20.3.11.97 --> 20.8.7.15`

* Non-default settings, if any
`allow_suspicious_codecs=1`

* Queries to run that lead to unexpected result
```sql
alter table test_table modify column index_column String CODEC(ZSTD(1))
```

**Error message and/or stacktrace**
`Code: 44, e.displayText() = DB::Exception: Trying to ALTER key column index_column (version 20.3.11.97 (official build))`

**Workaround**
drop index and modify column works
```sql
alter table test_table drop index index_name;
alter table test_table modify column index_column String CODEC(ZSTD(1));
```

**Comment by alexey-milovidov** [comment]
This is unlikely to be fixed.

**Comment by siradjev** [comment]
> This is unlikely to be fixed.

Recreating index on >100TB data is a very long-time consuming operation, making "no-impact" upgrade on live large cluster virtually impossible. 
We'll either have to slow down select queries, or arrange downtime on cluster to do manual change of SQL files in metadata and zookeeper.

**Comment by filimonov** [comment]
Hard to follow the sequence.

Did you get that problem during the update to 20.8 or during `alter table test_table modify column index_column String CODEC(ZSTD(1)) `

The alter was still running during update? Were you doing those alters on 20.3 or on 20.8?

Can you explain the scenario step by step? Initial schema on 20.3 (was it working?), sequence of alters and version upgrade.

**Comment by den-crane** [comment]
> Gorilla on LowCardinality column is an extremely weird idea.

Earlier CH allowed unusable CODECS. I see the third report about upgrade problem because of that.
It easy to fix, need to ALTER CODEC before an upgrade.

The real issue that it's impossible to ALTER CODEC for a column with SKIP index. Need to DROP INDEX. ALTER CODEC. ADD INDEX.

But it seems already solved.

```sql
20.13.1.5273

drop table SKIP;
create table SKIP( A Int64, D Date) Engine=MergeTree order by A;
insert into SKIP select number, toDate('2020-10-10') + number%10000 from numbers(1000000);
alter table SKIP add index ix1 (D) TYPE minmax GRANULARITY 2;
alter table SKIP modify column D CODEC(Delta, ZSTD);
alter table SKIP materialize index ix1;
alter table SKIP modify column D CODEC(ZSTD);
```

**Comment by siradjev** [comment]
>alter table test_table modify column index_column String CODEC(ZSTD(1))
This statement on 20.3 gives error. 


> Initial schema on 20.3 (was it working?) - Gorilla on LowCardinality column is an extremely weird idea.
It was initial design on 19.x was working. Was selected after test on all possible combinations on this column. 

> sequence of alters and version upgrade.
# upgrade, fail to start
# rollback
# trying to alter, failing.

**Comment by siradjev** [comment]
>The real issue that WAS

Issue is you have to modify index on 20.3.

**Comment by den-crane** [comment]
I think it's nothing to fix here. 

`LowCardinality(String) CODEC (Gorilla, ...)` probably can work because LowCardinality dictionary key has the same size over part.

**Comment by filimonov** [comment]
> > sequence of alters and version upgrade.
> 
>     * upgrade, fail to start
> 
>     * rollback
> 
>     * trying to alter, failing.


Why did you rollback instead of setting `allow_suspicious_codecs`?

**Comment by den-crane** [comment]
allow_suspicious_codecs -- for another cases like CODEC(ZSTD, ZSTD))

```
20.13.1

set allow_suspicious_codecs=1;

create table test(A LowCardinality(String) CODEC(Gorilla, ZSTD)) Engine=MergeTree order by tuple()
 DB::Exception: Codec Gorilla is not applicable for String because the data type is not of fixed size.
```

**Comment by siradjev** [comment]
> Why did you rollback instead of setting allow_suspicious_codecs?

it only solves below error. 

> SQL Error [36]: ClickHouse exception, code: 36, host: 127.0.0.1, port: 54669; Code: 36, e.displayText() = DB::Exception: It does not make sense to have codec NONE along with other compression codecs: NONE, ZSTD(1). (Note: you can enable setting 'allow_suspicious_codecs' to skip this check). (version 20.8.7.15 (official build))

**Comment by filimonov** [comment]
OK. I see, related to https://github.com/ClickHouse/ClickHouse/pull/12308/
